oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java

```diff
@@ -142,6 +214,74 @@ private static OIDCProviderMetadata parse(String body) {
 		}
 	}
 
+	private static void validateIssuer(OIDCProviderMetadata metadata, String issuer, ClientType type) {
+		String metadataIssuer = metadata.getIssuer().getValue();
+		if (!issuer.equals(metadataIssuer)) {
+			throw new IllegalStateException("The Issuer \"" + metadataIssuer + "\" provided in the " + getTypeName(type) +
+					" did not match the requested issuer \"" + issuer + "\"");
+		}
+	}
+
+	private static String getTypeName(ClientType type) {
+		return type.equals(ClientType.OIDC) ? "OpenID Configuration" : "OAuth Configuration";
+	}
+
+	/**
+	 * As per the <a href="https://tools.ietf.org/html/rfc8414#section-5">Section 5</a> the first attempt should check
+	 * https://example.com/.well-known/openid-configuration/issuer1 and if call failed then for backward
+	 * compatibility check https://example.com/issuer1/.well-known/openid-configuration when dealing with OpenID only.
+	 *
+	 * @param issuer
+	 * @param type
+	 * @param firstAttempt
+	 * @return
+	 */
+	private static String composeServerMetadataUrl(String issuer, ClientType type, boolean firstAttempt) {
+		String completePath = "";
+
+		/**
+		 * As per <a href="https://tools.ietf.org/html/rfc8414#section-3">Section 3</a> of RFC8414 the following
+		 * Creates "/.well-known/openid-configuration" OR "/.well-known/oauth-server-authorization"
+		 */
+		String wellknownPath = WELL_KNOWN_PATH + type.getIssuerIdentifier();
+
+		URI uri = URI.create(issuer);
+		String issuerPath = uri.getPath();
```

Comment:
I think a lot of this URL work can be cleaned up with `UriComponentsBuilder`.

Reply:
Thanks for suggesting `UriComponentBuilder` I have used that, let me know what you think of recent changes.